try to implement lockless SampleBank owned by output fn alone

acobster opened the pull request:
The main goal here is to avoid locks in audio callbacks. There are performance issues that could arise as well as logic issues as described [here](https://github.com/acobster/looper_proto/issues/4#issuecomment-885121599). Instead, we push each clip of samples over a channel. The output callback listens on the other end of the channel and is thus the sole owner of the samples vector.

Got stuck on:

```
   Compiling looper_proto v0.1.0 (/home/tamayo/projects/looper_proto)
error[E0525]: expected a closure that implements the `FnMut` trait, but this closure only implements `FnOnce`
   --> src/main.rs:84:26
    |
84  |     let output_data_fn = move |data: &mut [f32], _: &cpal::OutputCallbackInfo| {
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this closure implements `FnOnce`, not `FnMut`
...
96  |                     bank.write_at(clip.start, clip.samples);
    |                     ---- closure is `FnOnce` because it moves the variable `bank` out of its environment
...
137 |     let output_stream = output.build_output_stream(&config, output_data_fn, err_fn)?;
    |                                ------------------- the requirement to implement `FnMut` derives from here

error: aborting due to previous error
```

Commit:
try to implement lockless SampleBank owned by output fn alone

got stuck on:

error[E0525]: expected a closure that implements the `FnMut` trait, but this closure only implements `FnOnce`
   --> src/main.rs:85:26
    |
85  |     let output_data_fn = move |data: &mut [f32], _: &cpal::OutputCallbackInfo| {
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this closure implements `FnOnce`, not `FnMut`
...
97  |                     bank.write_at(clip.start, clip.samples);
    |                     ---- closure is `FnOnce` because it moves the variable `bank` out of its environment
...
139 |     let output_stream = output.build_output_stream(&config, output_data_fn, err_fn)?;
    |                                ------------------- the requirement to implement `FnMut` derives from here

Commit:
merge main -> lockless, fix conflicts manually

use both changesets

Commit:
rm unused import

btc commented:
@acobster strange. i've tried locally to let the output callback take bank as a reference but no luck so far.

I tried both of these and neither worked:

```
96 |                     (&bank).write_at(clip.start, clip.samples);
```

```
97 |                     bank.borrow().write_at(clip.start, clip.samples);
```

Commit:
retain ownership of bank in output_data_fn

Take self by reference to SampleBank::write_at(). This allows the caller
to retain ownership of the bank. Without this, the output_data_fn loses
ownership and thus ends up being a FnOnce rather than a FnMut, which cpal
expects.

Commit:
Move more state logic into output callback

Also:
* Fix incorrect total_samples logic (was using loop_len as a starting value
  instead of total_samples)
* Comment or remove println! stmts

acobster commented:
@btc Garrett helped me figure this out. The root cause had nothing to do with how `output_data_fn` was declared - the problem was actually `write_at`! See https://github.com/acobster/looper_proto/pull/5/commits/4f58027cbf6c1ffcee99a96df55bdf52fc3d440e.

btc commented:
@acobster wow subtle!
